src/mlpack/methods/kde/kde_rules.hpp

```diff
@@ -105,15 +123,41 @@ class KDERules
   //! Relatve error tolerance.
   const double relError;
 
+  //! Significance level for relative error compliance for Monte Carlo
+  //! estimations.
+  const double mcBeta;
+
+  //! Initial sample size for Monte Carlo estimations.
+  const size_t initialSampleSize;
+
+  //! Coefficient to control how much larger does the amount of node descendants
+  //! has to be compared to the initial sample size in order to be a candidate
+  //! for Monte Carlo estimations.
+  const double mcAccessCoef;
+
+  //! Coefficient to control what fraction of the amount of node's descendants
+  //! is the limit before Monte Carlo estimation recurses.
+  const double mcBreakCoef;
+
   //! Instantiated metric.
   MetricType& metric;
 
   //! Instantiated kernel.
   KernelType& kernel;
 
+  //! Whether Monte Carlo estimations are going to be applied.
+  const bool monteCarlo;
+
+  //! Accumulated not used MC alpha values for each query point.
+  arma::vec accumMCAlpha;
```

Comment:
I think that `alpha` here refers to the absolute error that has been used so far; maybe it would be better to give the variable a name closer than that or add some clarity to the comment?

Reply:
I agree it needs to be clarified. Before doing it, as far as I had understood, I think ```alpha``` refers to the probability of an MC estimation **not** being bounded by the relative error.

Maybe I misunderstood the "reclaiming probability" of [this paper](http://papers.nips.cc/paper/3539-fast-high-dimensional-kernel-summations-using-the-monte-carlo-multipole-method.pdf) but I think it doesn't refer to absolute error (although [the other paper](https://arxiv.org/pdf/1206.6857.pdf) does refer to absolute error).

This is what I have in mind: The bigger ```\alpha_{node}``` is, the more likely the scoring function is to use MC estimation. If some part of the tree is evaluated in any way that is not MC estimation, then its probability has not been used, therefore it can be used in the next scoring function evaluation. The sum of ```\alpha_{node}``` values at different nodes has to be equal to the original alpha input value of the rules (this means, the total probability of MC estimated nodes not being in the relative error bound is the same as the original ```alpha``` input value).

Did I miss or misunderstand anything?

Reply:
Ah, I see, thanks for putting the time into writing that up.  When I wrote my original comment, I did think what was being referred to was the absolute error reclamation of the "Faster Gaussian Summation: Theory and Experiment".  So, I see now that what you have written is right, and the only thing that remains will be to actually review the code in depth. :)  But it is getting late here (my evening was mostly taken up handling some around-the-house things), so I'll review the code tomorrow over breakfast. :+1:

It might be worth thinking about doing the absolute error reclamation later, but, up to you---I don't think that was in your proposal, so maybe more exciting to move on to other parts. :)  We can always open an issue for it and leave it for later or someone else interested to do.  Let me know what you think.

Reply:
I agree, it's probably a very good idea to work on the absolute error reclamation, but I think if we want to finish what was on the proposal, then we can open an issue and I can take care of this later :)

Reply:
Sounds good to me. :+1:

Reply:
Issue  #1954 open :)